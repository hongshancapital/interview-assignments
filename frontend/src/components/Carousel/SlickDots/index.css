.carousel-dots {
  position: absolute;
  width: 100%;
  bottom: 60px;
  display: flex;
  justify-content: center;
}
.carousel-dots > div {
  margin-left: 16px;
}
.carousel-dots > div:first-child {
  margin-left: 0;
}
.carousel-dots .progress {
  width: 50px;
}
