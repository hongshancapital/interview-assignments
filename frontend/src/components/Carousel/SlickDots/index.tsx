import { memo, FC } from 'react';
import Progress from '../../Progress';
import './index.css';

interface SlickDotsProps {
  /** number of panel indication points */
  length: number;
  /** currently selected index value */
  activeIndex: number;
  /** animated duration */
  duration?: number;
  /** click event */
  onClick?: (index: number) => void;
}

const SlickDots: FC<SlickDotsProps> = function ({
  length,
  activeIndex,
  duration = 1,
  onClick,
}) {
  const renderDot = () => {
    let list = [];
    for (let i = 0; i < length; i++) {
      const isActive = activeIndex === i;
      list.push(
        <Progress
          key={i}
          percent={isActive ? '100%' : '0%'}
          barStyle={isActive ? { transition: `${duration}s all` } : undefined}
          onClick={() => onClick?.(i)}
        />
      );
    }
    return list;
  };
  return <div className="carousel-dots">{renderDot()}</div>;
};

export default memo(SlickDots);
