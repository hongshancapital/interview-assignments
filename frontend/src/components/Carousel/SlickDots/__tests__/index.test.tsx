import { render, act } from '@testing-library/react';
import SlickDots from '../index';
describe('test Progress component', () => {
  it('renders correctly', () => {
    const { container } = render(<SlickDots length={4} activeIndex={0} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('mount correctly', () => {
    expect(() =>
      render(<SlickDots length={4} activeIndex={0} />)
    ).not.toThrow();
  });

  it('test rendering properties', () => {
    const { container } = render(
      <SlickDots length={4} activeIndex={0} duration={3} />
    );

    expect(container.firstChild?.childNodes).toHaveLength(4);

    expect(
      container.querySelectorAll('.carousel-dots .progress-bar')[0]
    ).toHaveStyle({ transition: '3s all', width: '100%' });

    expect(
      container.querySelectorAll('.carousel-dots .progress-bar')[1]
    ).not.toHaveStyle({ transition: '3s all' });

    expect(
      container.querySelectorAll('.carousel-dots .progress-bar')[1]
    ).toHaveStyle({ width: '0%' });
  });

  it('test click event', () => {
    const onClickItem = jest.fn();

    const { container } = render(
      <SlickDots
        length={4}
        activeIndex={0}
        duration={3}
        onClick={onClickItem}
      />
    );
    const dot = container.querySelectorAll('.carousel-dots .progress-bar')[2];

    act(() => {
      dot.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    });

    expect(onClickItem).toBeCalledTimes(1);
    expect(onClickItem).toBeCalledWith(2);

    act(() => {
      for (let i = 0; i < 5; i++) {
        dot.dispatchEvent(new MouseEvent('click', { bubbles: true }));
      }
    });

    expect(onClickItem).toHaveBeenCalledTimes(6);
  });
});
