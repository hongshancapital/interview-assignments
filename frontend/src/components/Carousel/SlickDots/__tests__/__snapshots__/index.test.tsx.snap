// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress renders correctly 1`] = `
<div
  class="carousel-dots"
>
  <div
    class="progress "
  >
    <div
      class="progress-bar"
      style="width: 100%; transition: 1s all;"
    />
  </div>
  <div
    class="progress "
  >
    <div
      class="progress-bar"
      style="width: 0%;"
    />
  </div>
  <div
    class="progress "
  >
    <div
      class="progress-bar"
      style="width: 0%;"
    />
  </div>
  <div
    class="progress "
  >
    <div
      class="progress-bar"
      style="width: 0%;"
    />
  </div>
</div>
`;

exports[`test Progress component renders correctly 1`] = `
<div
  class="carousel-dots"
>
  <div
    class="progress "
  >
    <div
      class="progress-bar"
      style="width: 100%; transition: 1s all;"
    />
  </div>
  <div
    class="progress "
  >
    <div
      class="progress-bar"
      style="width: 0%;"
    />
  </div>
  <div
    class="progress "
  >
    <div
      class="progress-bar"
      style="width: 0%;"
    />
  </div>
  <div
    class="progress "
  >
    <div
      class="progress-bar"
      style="width: 0%;"
    />
  </div>
</div>
`;
